src/components/Tables/Table.tsx

```diff
@@ -0,0 +1,149 @@
+    for (const key of Object.keys(headers)) {
+        if (
+            !(item[key] ?? "")
+                .toString()
+                .toLowerCase()
+                .includes((filterText[key] ?? "").toLowerCase())
+        ) {
+            return false;
+        }
+    }
+    return true;
+};
+
+const filteredNoNullItems = (headers: Headers, data: Datum[], filterText: FilterText): Row[] => {
+    let row = 0;
+    return data.reduce((filteredNoNullList: Row[], item: Datum, currentIndex: number) => {
+        if (itemIncludesFilterText(headers, item, filterText)) {
+            const noNullItem: Row = { checkBoxKey: currentIndex };
+            for (const key of Object.keys(headers)) {
+                noNullItem[key] = item[key] ?? "";
+            }
+            noNullItem.id = row;
+            filteredNoNullList.push(noNullItem);
+            row++;
+        }
+        return filteredNoNullList;
+    }, []);
+};
+
+const Table: React.FC<Props> = (props) => {
+    const [filterText, setFilterText] = useState<FilterText>({});
+
+    const [selectCheckBox, setSelectCheckBox] = useState(
+        new Array<boolean>(props.data.length).fill(false)
+    );
+    const [selectAllCheckBox, setSelectAllCheckBox] = useState(false);
+
+    const toggleAllCheckBox = (): void => {
+        setSelectCheckBox(new Array<boolean>(props.data.length).fill(!selectAllCheckBox));
+        setSelectAllCheckBox(!selectAllCheckBox);
+    };
+    const toggleOwnCheckBox = (checkBoxKey: number): void => {
+        const selectCheckBoxCopy = [...selectCheckBox];
+        selectCheckBoxCopy[checkBoxKey] = !selectCheckBoxCopy[checkBoxKey];
+        setSelectCheckBox(selectCheckBoxCopy);
+    };
+
+    // Fixes hydration error by re-rendering when the DOM is ready - https://nextjs.org/docs/messages/react-hydration-error
+    const [domLoaded, setDomLoaded] = useState(false);
+    useEffect(() => {
+        setDomLoaded(true);
+    }, []);
+
+    useEffect(() => {
+        let toggleFlag = true;
+        selectCheckBox.forEach((eachCheckBox) => {
+            if (!eachCheckBox) {
+                toggleFlag = false;
+            }
+        });
+        setSelectAllCheckBox(toggleFlag);
+    }, [selectCheckBox]);
+
+    const columns: TableColumn<Row>[] = Object.entries(props.headers).map(([key, value]) => {
+        return {
+            name: value,
+            selector: (row: Row) => row[key],
+            sortable: true,
+        };
+    });
+
+    columns.unshift({
+        name: (
+            <input
+                type="checkbox"
+                aria-label="Select all rows"
+                checked={selectAllCheckBox}
+                onClick={toggleAllCheckBox}
+            />
+        ),
+        cell: (row: Row) => (
+            <input
+                type="checkbox"
+                aria-label={`Select row ${row.checkBoxKey}`}
+                checked={selectCheckBox[Number(row.checkBoxKey)]}
```

Comment:
why are these `Number(...)` wrapping them? they should just be good?

Reply:
It was complaining just now but it didn't complain now (probably because at first we mistakenly used row.id) -> so refactored as suggested